inc/c_util/thandle_tuple_array.h

```diff
@@ -0,0 +1,147 @@
+                required_size = count * sizeof(name); \
+                result->tuple_array = (name *)malloc(required_size); \
+                \
+                if (result->tuple_array == NULL) \
+                { \
+                    /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_004: [ If there are any errors then THANDLE_TUPLE_ARRAY_CREATE(name) shall fail and return NULL. ]*/ \
+                    LogError("malloc(%zu) failed for count=%" PRIu32, required_size, count); \
+                    failed = true; \
+                } \
+            } \
+            \
+            if (failed) \
+            { \
+                /* already logged */ \
+            } \
+            else \
+            { \
+                /* Need to cast to assign to the const member */ \
+                *(uint32_t*)&result->count = count; \
+                \
+                /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_003: [ THANDLE_TUPLE_ARRAY_CREATE(name) shall initialize the members of the tuples in the array to NULL. ]*/ \
+                for (uint32_t i = 0; i < count; ++i) \
+                { \
+                    MU_FOR_EACH_2(THANDLE_TUPLE_ARRAY_INIT_MEMBER, __VA_ARGS__) \
+                } \
+                \
+                /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_005: [ THANDLE_TUPLE_ARRAY_CREATE(name) shall succeed and return the allocated array. ]*/ \
+                goto all_ok; \
+            } \
+            free(result); \
+            result = NULL; \
+        } \
+    all_ok: \
+        return result; \
+    }
+
+#define THANDLE_TUPLE_ARRAY_FREE_MEMBER(type, name) \
+    if (tuple_array->tuple_array[i].name != NULL) \
+    { \
+        THANDLE_DEC_REF(type)(tuple_array->tuple_array[i].name); \
+    }
+
+#define IMPLEMENT_THANDLE_TUPLE_ARRAY_DESTROY(name, ...) \
+    IMPLEMENT_MOCKABLE_FUNCTION(, void, THANDLE_TUPLE_ARRAY_DESTROY(name), THANDLE_TUPLE_ARRAY_TYPE(name)*, tuple_array) \
+    { \
+        if (tuple_array == NULL) \
+        { \
+            /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_006: [ If tuple_array is NULL then THANDLE_TUPLE_ARRAY_DESTROY(name) shall fail and return. ]*/ \
+            LogError("Invalid args: " MU_TOSTRING(THANDLE_TUPLE_ARRAY_TYPE(name)) "* tuple_array = %p", \
+                tuple_array); \
+        } \
+        else \
+        { \
+            /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_007: [ THANDLE_TUPLE_ARRAY_DESTROY(name) shall iterate over all of the elements in tuple_array and call THANDLE_DEC_REF(type) for each field. ]*/ \
+            for (uint32_t i = 0; i < tuple_array->count; ++i) \
+            { \
+                MU_FOR_EACH_2(THANDLE_TUPLE_ARRAY_FREE_MEMBER, __VA_ARGS__) \
+            } \
+            /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_008: [ THANDLE_TUPLE_ARRAY_DESTROY(name) shall free the memory allocated in tuple_array. ]*/ \
+            if (tuple_array->tuple_array != NULL) \
+            { \
+                free((void*)tuple_array->tuple_array); \

```

Comment:
>(void*) [](http://example.com/codeflow?start=21&length=7)

what happens if you remove this cast?

Reply:
can't free a const pointer (and that's good because it catches mistakes of calling free instead of destroy)